x/distribution/keeper/allocation.go

```diff
@@ -51,35 +51,62 @@ func (k Keeper) AllocateTokens(ctx context.Context, totalPreviousPower int64, bo
 	voteMultiplier := math.LegacyOneDec().Sub(communityTax)
 	feeMultiplier := feesCollected.MulDecTruncate(voteMultiplier)
 
-	// allocate tokens proportionally to voting power
-	//
-	// TODO: Consider parallelizing later
-	//
-	// Ref: https://github.com/cosmos/cosmos-sdk/pull/3099#discussion_r246276376
-	for _, vote := range bondedVotes {
+	var wg sync.WaitGroup
+	// channel for errors
+	errc := make(chan error, len(bondedVotes))
+	// channel for remaining rewards
+	remainingc := make(chan sdk.DecCoins, len(bondedVotes))
 
-		validator, err := k.stakingKeeper.ValidatorByConsAddr(ctx, vote.Validator.Address)
-		if err != nil {
-			return err
-		}
+	for _, vote := range bondedVotes {
+		wg.Add(1)
+		go func(vote comet.VoteInfo) {
+			defer wg.Done()
+
+			validator, err := k.stakingKeeper.ValidatorByConsAddr(ctx, vote.Validator.Address)
+			if err != nil {
+				errc <- err
+				return
+			}
+
+			powerFraction := math.LegacyNewDec(vote.Validator.Power).QuoTruncate(math.LegacyNewDec(totalPreviousPower))
+			reward := feeMultiplier.MulDecTruncate(powerFraction)
+
+			if err = k.AllocateTokensToValidator(ctx, validator, reward); err != nil {
+				errc <- err
+				return
+			}
+
+			remainingc <- reward
+		}(vote)
+	}
+	wg.Wait()
 
-		// TODO: Consider micro-slashing for missing votes.
-		//
-		// Ref: https://github.com/cosmos/cosmos-sdk/issues/2525#issuecomment-430838701
-		powerFraction := math.LegacyNewDec(vote.Validator.Power).QuoTruncate(math.LegacyNewDec(totalPreviousPower))
-		reward := feeMultiplier.MulDecTruncate(powerFraction)
+	// Close the channels
+	close(errc)
+	close(remainingc)
```

Comment:
I highly recommend putting the wg.Wait() and channel closing in a goroutine so that you can process data concurrently otherwise here you'll just be waiting for channels to be fully buffered after all the data has processed which defeats the original purpose. In fact I'd even recommend deriving a context that's cancellable and on any error, invoke cancel() so that any future processing inside bondedVotes will immediately cease so
```go
        ctx, cancel := context.WithCancel(ctx)
        defer cancel()

        for _, vote := range bondedVotes {
		wg.Add(1)
		go func(vote comet.VoteInfo) {
			defer wg.Done()

                         if err := ctx.Err(); err != nil { // Context cancellation, return immediately, no need to send this error
                                return
                         }

			validator, err := k.stakingKeeper.ValidatorByConsAddr(ctx, vote.Validator.Address)
			if err != nil {
				errCh <- err
				return
			}

			powerFraction := math.LegacyNewDec(vote.Validator.Power).QuoTruncate(math.LegacyNewDec(totalPreviousPower))
			reward := feeMultiplier.MulDecTruncate(powerFraction)

			if err = k.AllocateTokensToValidator(ctx, validator, reward); err != nil {
				errCh <- err
				return
			}

			remainingc <- reward
		}(vote)
	}

        go func() {
	     wg.Wait()
             close(errCh)
             close(remainingCh)
        }()

        for err := range errCh {
                if err != nil {
                        return err
                }
        }
```